feat(net): Prevent removing networks that are in use

LucaSeri opened the pull request:
Check whether there are any running machines that are using the network before removing it. Additionally introduce a force flag that can override this.

<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes
This PR adds a check that prevents a user from removing a network in use, unless forcing it.

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Commit:
feat(net): Prevent removing networks that are in use

Check whether there are any running machines that are
using the network before removing it. Additionally
introduce a force flag that can override this.

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

craciunoiuc reviewed (commented):
looks good on an initial view

once out of draft I'll also try it out

jake-ciolek commented on internal/cli/kraft/net/remove/remove.go:
```diff
@@ -65,6 +68,29 @@ func (opts *RemoveOptions) Run(ctx context.Context, args []string) error {
+			continue
+		}
+		for _, network := range machine.Spec.Networks {
+			if network.IfName == args[0] {
```
It'd be good to ensure the length of args somewhere before this.

LucaSeri replied:
Nice catch, thanks for this! I added an additional check.

jake-ciolek commented on internal/cli/kraft/net/remove/remove.go:
```diff
@@ -65,6 +68,29 @@ func (opts *RemoveOptions) Run(ctx context.Context, args []string) error {
+		for _, network := range machine.Spec.Networks {
+			if network.IfName == args[0] {
+				if !opts.Force {
+					return fmt.Errorf("network %s is in use by machine %s. Use --force to remove it anyway", args[0], machine.Name)
```
see above

Commit:
fix(net): Check argument length

When using kraftkit as an API, the CLI check
on the number of arguments is not done, resulting
in potentially calling kraft net rm with no arguments.

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

nderjung commented on internal/cli/kraft/net/remove/remove.go:
```diff
@@ -65,6 +72,42 @@ func (opts *RemoveOptions) Run(ctx context.Context, args []string) error {
+				if !opts.Force {
+					return fmt.Errorf("network %s is in use by machine %s. Use --force to remove it anyway", args[0], machine.Name)
+				} else {
+					fmt.Fprintf(iostreams.G(ctx).ErrOut, "warning: network %s is in use by machine %s. Removing anyway\n", args[0], machine.Name)
```
Let's use the native warning log-level:
```go
log.G(ctx).Warnf("network '%s' is in use by machine '%s'", args[0], machine.Name)
```

LucaSeri replied:
Fixed, thanks!

nderjung reviewed (approved):
Thanks!
